File: lib/app/services/keys_services/keys_service.dart
import 'dart:convert';
import 'dart:math';
import 'dart:typed_data';

import 'package:flutter/foundation.dart';
import 'package:pointycastle/api.dart';
import 'package:pointycastle/asymmetric/api.dart';
import 'package:pointycastle/digests/sha256.dart';
import 'package:pointycastle/key_generators/api.dart';
import 'package:pointycastle/key_generators/rsa_key_generator.dart';
import 'package:pointycastle/random/fortuna_random.dart';
import 'package:pointycastle/signers/rsa_signer.dart';

class KeysServices {
  static const String _pubHeader = "-----BEGIN NEIGHBOR_NEWS PUBLIC KEY-----";
  static const String _pubFooter = "-----END NEIGHBOR_NEWS PUBLIC KEY-----";

  static const String _privHeader = "-----BEGIN NEIGHBOR_NEWS PRIVATE KEY-----";
  static const String _privFooter = "-----END NEIGHBOR_NEWS PRIVATE KEY-----";

  static AsymmetricKeyPair<RSAPublicKey, RSAPrivateKey> generateKeyPair() {
    final secureRandom = FortunaRandom();
    final random = Random.secure();
    final seeds = List<int>.generate(32, (_) => random.nextInt(256));
    secureRandom.seed(KeyParameter(Uint8List.fromList(seeds)));

    final rsaParams = RSAKeyGeneratorParameters(BigInt.from(65537), 2048, 64);
    final params = ParametersWithRandom(rsaParams, secureRandom);

    final keyGenerator = RSAKeyGenerator();
    keyGenerator.init(params);

    final pair = keyGenerator.generateKeyPair();
    debugPrint("Chave Publica: ${pair.publicKey.toString()}");
    debugPrint("Chave Privada: ${pair.privateKey}");
    return AsymmetricKeyPair<RSAPublicKey, RSAPrivateKey>(
      pair.publicKey,
      pair.privateKey,
    );
  }

  static String exportPublicKey(RSAPublicKey key) {
    final modulus = key.modulus.toString();
    final exponent = key.exponent.toString();
    final rawString = "$modulus|$exponent";

    final base64Key = base64.encode(utf8.encode(rawString));
    debugPrint("Chave Publica em string: $base64Key");
    return "$_pubHeader\n$base64Key\n$_pubFooter";
  }

  static String exportPrivateKey(RSAPrivateKey key) {
    final modulus = key.modulus.toString();
    final privateExponent = key.privateExponent.toString();
    final rawString = "$modulus|$privateExponent";

    final base64Key = base64.encode(utf8.encode(rawString));
    debugPrint("Chave Publica em string: $base64Key");
    return "$_privHeader\n$base64Key\n$_privFooter";
  }

  static RSAPublicKey importPublicKey(String pemString) {
    if (!pemString.contains(_pubHeader) || !pemString.contains(_pubFooter)) {
      throw FormatException(
          "Chave inválida! Esta chave não pertence ao projeto Neighbor News.");
    }

    final base64Content = pemString
        .replaceAll(_pubHeader, "")
        .replaceAll(_pubFooter, "")
        .replaceAll("\n", "")
        .trim();

    final rawString = utf8.decode(base64.decode(base64Content));
    final parts = rawString.split('|');

    final modulus = BigInt.parse(parts[0]);
    final exponent = BigInt.parse(parts[1]);

    return RSAPublicKey(modulus, exponent);
  }

  static RSAPrivateKey importPrivateKey(String pemString) {
    if (!pemString.contains(_privHeader) || !pemString.contains(_privFooter)) {
      throw FormatException(
          "Chave inválida! Esta chave não pertence ao projeto Neighbor News.");
    }

    final base64Content = pemString
        .replaceAll(_privHeader, "")
        .replaceAll(_privFooter, "")
        .replaceAll("\n", "")
        .trim();

    final rawString = utf8.decode(base64.decode(base64Content));
    final parts = rawString.split('|');

    final modulus = BigInt.parse(parts[0]);
    final privateExponent = BigInt.parse(parts[1]);

    return RSAPrivateKey(modulus, privateExponent, null, null);
  }

  static String toSign(String newsBody, RSAPrivateKey privateKey) {
    final signer =
        RSASigner(SHA256Digest(), '0609608648016503040201'); // OID para SHA-256
    signer.init(true, PrivateKeyParameter<RSAPrivateKey>(privateKey));

    final bytesToSign = utf8.encode(newsBody);
    final signature = signer.generateSignature(Uint8List.fromList(bytesToSign));
    debugPrint("Assinatura: ${base64.encode(signature.bytes)}");
    return base64.encode(signature.bytes);
  }

  static bool toCheck(
      String newsBody, String assinaturaBase64, RSAPublicKey publicKey) {
    try {
      final verifier = RSASigner(SHA256Digest(), '0609608648016503040201');
      verifier.init(false, PublicKeyParameter<RSAPublicKey>(publicKey));

      final bytesToVerify = utf8.encode(newsBody);
      final signatureBytes = base64.decode(assinaturaBase64);
      final signature = RSASignature(Uint8List.fromList(signatureBytes));

      return verifier.verifySignature(
          Uint8List.fromList(bytesToVerify), signature);
    } catch (e) {
      return false;
    }
  }
}

File: functions/src/service/keys-service.ts
import * as forge from 'node-forge';

const PUB_HEADER = "-----BEGIN NEIGHBOR_NEWS PUBLIC KEY-----";
const PUB_FOOTER = "-----END NEIGHBOR_NEWS PUBLIC KEY-----";

const PRIV_HEADER = "-----BEGIN NEIGHBOR_NEWS PRIVATE KEY-----";
const PRIV_FOOTER = "-----END NEIGHBOR_NEWS PRIVATE KEY-----";

export const KeysService = {

  generateKeyPair: () => {
    const { publicKey, privateKey } = forge.pki.rsa.generateKeyPair({ bits: 2048, e: 0x10001 });
    return { publicKey, privateKey };
  },

  exportPublicKey: (publicKey: forge.pki.rsa.PublicKey): string => {
    const modulus = publicKey.n.toString(10);
    const exponent = publicKey.e.toString(10);

    const rawString = `${modulus}|${exponent}`;
    const base64Key = Buffer.from(rawString, 'utf8').toString('base64');
    return `${PUB_HEADER}\n${base64Key}\n${PUB_FOOTER}`;
  },

  exportPrivateKey: (privateKey: forge.pki.rsa.PrivateKey): string => {
    const modulus = privateKey.n.toString(10);
    const privateExponent = privateKey.d.toString(10);

    const rawString = `${modulus}|${privateExponent}`;
    const base64Key = Buffer.from(rawString, 'utf8').toString('base64');
    return `${PRIV_HEADER}\n${base64Key}\n${PRIV_FOOTER}`;
  },

  importPublicKey: (pemString: string): forge.pki.rsa.PublicKey => {
    if (!pemString.includes(PUB_HEADER) || !pemString.includes(PUB_FOOTER)) {
      throw new Error("Invalid key");
    }

    const base64Content = pemString
      .replace(PUB_HEADER, "")
      .replace(PUB_FOOTER, "")
      .replace(/\n/g, "")
      .trim();

    const rawString = Buffer.from(base64Content, 'base64').toString('utf8');
    const parts = rawString.split('|');
    const modulus = parts[0];
    const exponent = parts[1];

    const n = new forge.jsbn.BigInteger(modulus, 10);
    const e = new forge.jsbn.BigInteger(exponent, 10);

    return forge.pki.rsa.setPublicKey(n, e);
  },

  importPrivateKey: (pemString: string): forge.pki.rsa.PrivateKey => {
    if (!pemString.includes(PRIV_HEADER) || !pemString.includes(PRIV_FOOTER)) {
      throw new Error("Invalid key");
    }

    const base64Content = pemString
      .replace(PRIV_HEADER, "")
      .replace(PRIV_FOOTER, "")
      .replace(/\n/g, "")
      .trim();

    const rawString = Buffer.from(base64Content, 'base64').toString('utf8');
    const parts = rawString.split('|');
    const modulus = parts[0];
    const privateExponent = parts[1];

    const n = new forge.jsbn.BigInteger(modulus, 10);
    const d = new forge.jsbn.BigInteger(privateExponent, 10);
    const e = new forge.jsbn.BigInteger('65537', 10);

    // @ts-ignore
    return forge.pki.rsa.setPrivateKey(n, e, d);
  },

  toSign: (payload: string, privateKey: forge.pki.rsa.PrivateKey): string => {
    const md = forge.md.sha256.create();
    md.update(payload, 'utf8');
    const signature = privateKey.sign(md);
    return forge.util.encode64(signature);
  },

  toCheck: (payload: string, signature: string, publicKey: forge.pki.rsa.PublicKey): boolean => {
    try {
      const decodedSignature = forge.util.decode64(signature);
      const md = forge.md.sha256.create();
      md.update(payload, 'utf8');
      return publicKey.verify(md.digest().bytes(), decodedSignature);
    } catch (e) {
      return false;
    }
  }
};
